### DISCUSSION TITLE
Circular Scrolling List Version 5

### DISCUSSION THREAD

**Discussion by LanKuDot** [LanKuDot/game_modules#17] [Ideas]
**Version 5 is completed!** 🎉 

Here are the new features or improvements to be added to the next update:

- [x] Reverse order option ([Quick patch for ver 4.](https://gist.github.com/LanKuDot/7b56a6d2a7415cc4f65419ff1f3a2a58))
- [x] Move the clicked box to the center
- [x] Select the box from the script
- [x] Refresh the list contents at runtime
- [x] Custom ListBox (by inheriting the base class) (Issue #13)
- [x] Custom list area (Issue #9)
- [x] The setup class will be `CircularScrollingList` not `ListPositionCtrl`
- [x] Callback events
- [x] Image sorting - center in front
- [x] Code refactoring - Easy to develop

The code refactoring will take a lot of time, and I will do it first. (Such as change tabs to spaces, add namespace, and VisualStudio style comments).

The developing branch is `circularlist-dev`, and the progress is also recorded in [project](https://github.com/LanKuDot/game_modules/projects/1).

**Comment by Mykey05** [comment]
Hi. Is there a quick solution to making the Horizontal List go from Left to Right?

**Reply by LanKuDot** [reply]
The easiest way is just reverse the order of the contents provided from the `ListBank.cs` .
The other way is to modify the `ListBox.cs` as below, but it only works in the circular mode:
```diff
diff --git a/CircularScrollingList/ListBox.cs b/CircularScrollingList/ListBox.cs
index 98f22da..3e247b6 100644
--- a/CircularScrollingList/ListBox.cs
+++ b/CircularScrollingList/ListBox.cs
@@ -272,7 +272,7 @@ public class ListBox : MonoBehaviour
         _contentID = _positionCtrl.centeredContentID;

         // Adjust the contentID according to its initial order.
-        _contentID += listBoxID - _positionCtrl.listBoxes.Length / 2;
+        _contentID += _positionCtrl.listBoxes.Length / 2 - listBoxID;

         // In the linear mode, disable the box if needed
         if (_positionCtrl.listType == ListPositionCtrl.ListType.Linear) {
@@ -312,8 +312,8 @@ public class ListBox : MonoBehaviour
     /// </summary>
     private void UpdateToLastContent()
     {
-        _contentID = nextListBox.GetContentID() - 1;
-        _contentID = (_contentID < 0) ? _listBank.GetListLength() - 1 : _contentID;
+        _contentID = nextListBox.GetContentID() + 1;
+        _contentID = (int)Mathf.Repeat(_contentID, _listBank.GetListLength());

         if (_positionCtrl.listType == ListPositionCtrl.ListType.Linear) {
             if (_contentID == _listBank.GetListLength() - 1 ||
@@ -342,8 +342,8 @@ public class ListBox : MonoBehaviour
     /// </summary>
     private void UpdateToNextContent()
     {
-        _contentID = lastListBox.GetContentID() + 1;
-        _contentID = (_contentID == _listBank.GetListLength()) ? 0 : _contentID;
+        _contentID = lastListBox.GetContentID() - 1;
+        _contentID = (int)Mathf.Repeat(_contentID, _listBank.GetListLength());

         if (_positionCtrl.listType == ListPositionCtrl.ListType.Linear) {
             if (_contentID == 0 || !lastListBox.isActiveAndEnabled) {

```

**Reply by Mykey05** [reply]
Cheers, I did consider reversing the order but I wanted to keep it tidy for Inspector because in my project I am changing the contents dynamically. Anyways I stuck with having the contents list reversed and that works fine for a Linear List. Please consider including another method in BaseListBank that returns the reversed contentID if the user specified reversing the List. Awesome work!

**Reply by LanKuDot** [reply]
Thanks! I will solve this problem first, and give a patch to provide this function.

**Reply by LanKuDot** [reply]
Hi! @Mykey05 I made a quick patch that makes the list support reverse order in linear mode.
Check the patch here: https://gist.github.com/LanKuDot/7b56a6d2a7415cc4f65419ff1f3a2a58

**Reply by Mykey05** [reply]
Hi! Been working on other stuff lately and I haven't updated those lists yet with your patch but thanks a lot for updating it!

**Comment by weihsiangleow** [comment]
Hi, just wanted to check if there is a way to have multiple control mode at the same time? i want the list to be draggable and at the same time, if i click on like say an arrowup/arrowdown button, it will scroll the list using the the function MoveOneUnitUp/MoveOneUnitDown

**Reply by LanKuDot** [reply]
I though it could. You have to create another `MovementCtrl` and a moving curve for it, such as `FreeMovementCtrl` for the dragging, and `UnitMovementCtrl` for buttons. And use the latter one in the `SetUnitMove()`. Check both `movementCtrl` in the `Update()`.

**Reply by weihsiangleow** [reply]
Thanks, i got it working, really appreciate the work done here.
